### PR TITLE
feat (batcher): finalize batch every n blocks

### LINKED ISSUE #185: feat (batcher): finalize batch every N blocks
Batcher should finalize batch every N blocks or M proofs (whatever comes first) instead of hardcoded 2 proofs.

This should be configurable via config file.

### PR DESCRIPTION
**Pull request by taturosati**
https://github.com/yetanotherco/aligned_layer/pull/182 needs to be merged first

Batcher now finalizes batch (uploads to s3 & submits task to Ethereum every N blocks or M tasks, whichever happens first.

You can configure batch_size_interval & block_interval on config.yaml

# To test

1. Start anvil, aggregator, operator & batcher (make sure to have env file with aws credentials)
2. Run `make batcher_send_dummy_task` 5 times
3. You should see batcher finalizing batch
4. Run `make batcher_send_dummy_task` twice (lambdaworks Merkle tree requires at least two leaves)
5. Run `make operator_get_eth` (just to make anvil generate new blocks)
6. Batcher should finalize batch
7. You'll see a call reverted error because Merkle tree is repeated, you can test both succesfully by restarting anvil between batches

### CODE REVIEW

File: batcher/src/lib.rs [hunk 1]
```diff
@@ -81,13 +67,42 @@ impl App {
...
+            batch_size_interval: config.batcher.batch_size_interval,
+            last_uploaded_batch_block: Mutex::new(0),
         }
     }
 
-    pub async fn handle_connection(&self, raw_stream: TcpStream, addr: SocketAddr) {
+    pub async fn listen(self: &Arc<Self>, address: &str) {
+        // Create the event loop and TCP listener we'll accept connections on.
+        let listener = TcpListener::bind(address).await.expect("Failed to build");
+        // let listener = try_socket.expect("Failed to bind");
```
Comment (JuArce) [inline batcher/src/lib.rs]: remove this comment

**Review by JuArce — CHANGES_REQUESTED** [review]

File: batcher/src/lib.rs (hunk 1 above)
Comment (taturosati) [reply batcher/src/lib.rs]: removed on 21dfe2ed26c7bad3231a2c9bb95b3f9203ac9b22

**Review by JuArce — APPROVED** [review]
lgtm

File: batcher/src/lib.rs [hunk 2]
```diff
@@ -170,42 +183,75 @@ impl App {
...
-        let batch_merkle_root_hex = hex::encode(batch_merkle_tree.root);
-        info!("Batch merkle root: {}", batch_merkle_root_hex);
+    async fn handle_new_block(&self, block_number: u64) {
+        let (batch_bytes, batch_merkle_root) = {
+            let mut last_uploaded_batch_block = self.last_uploaded_batch_block.lock().await;
+            let mut current_batch = self.current_batch.lock().await;
+            let current_batch_len = current_batch.len();
+            if current_batch_len <= 1 {
+                // Needed because merkle tree freezes on only one leaf
+                info!("New Block reached but current batch is empty or has only one task");
```
Comment (MauroToscano) [inline batcher/src/lib.rs]: This shouldn't be happening

Comment (MauroToscano) [reply batcher/src/lib.rs]: We can fix it in another PR but we should make an issue

**Review by jotabulacios — APPROVED** [review]
lgtm

File: batcher/src/lib.rs (hunk 2 above)
Comment (taturosati) [reply batcher/src/lib.rs]: created issue https://github.com/yetanotherco/aligned_layer/issues/190

**Review by entropidelic — APPROVED** [review]

**Pull request merged**